ShadowContext中unregisterReceivery可能有问题

``fix(core.runtime): BroadcastReceiver与BroadcastReceiverWapper映射Map采用弱引用实现
这次的提交在反注册的时候应该是有问题的
起因是我这里报了一个异常Receiver not registered: com.tencent.shadow.core.runtime.BroadcastReceiverWrapper
然后看代码，反注册的时候未在mReceiverWrapperMap中的的还会往mReceiverWrapperMap中put一个，且反注册的时候没有remove掉相应的Receiver

![image](https://user-images.githubusercontent.com/13994354/159836355-31a893c5-d1c3-4aa1-b4bc-8a96a8f5e14e.png)

> 我猜应该是这个错误导致的：#867 可以测一下能不能复现问题了。

直接运行sample-host不崩吗？ 我这运行sample-host然后点启动插件就崩了

2022-03-24 15:31:18.617 12176-12230/com.tencent.shadow.sample.host E/AndroidRuntime: FATAL EXCEPTION: pool-3-thread-1
    Process: com.tencent.shadow.sample.host, PID: 12176
    java.lang.RuntimeException: android.os.DeadObjectException
        at com.tencent.shadow.sample.manager.SamplePluginManager$1.run(SamplePluginManager.java:142)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: android.os.DeadObjectException
        at android.os.BinderProxy.transactNative(Native Method)
        at android.os.BinderProxy.transact(Binder.java:503)
        at com.tencent.shadow.dynamic.manager.BinderPluginLoader.callApplicationOnCreate(BinderPluginLoader.java:79)
        at com.tencent.shadow.sample.manager.FastPluginManager.callApplicationOnCreate(FastPluginManager.java:121)
        at com.tencent.shadow.sample.manager.SamplePluginManager$1.run(SamplePluginManager.java:128)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
        at java.lang.Thread.run(Thread.java:818) 

--- a/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/blocs/CreateApplicationBloc.kt
+++ b/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/blocs/CreateApplicationBloc.kt
@@ -60,6 +60,7 @@ object CreateApplicationBloc {
             shadowApplication.applicationInfo = pluginApplicationInfo
             shadowApplication.setBusinessName(loadParameters.businessName)
             shadowApplication.setPluginPartKey(partKey)
+            shadowApplication.setShadowApplication(shadowApplication)
 
             //和ShadowActivityDelegate.initPluginActivity一样，attachBaseContext放到最后
             shadowApplication.setHostApplicationContextAsBase(hostAppContext)

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowApplication.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowApplication.java
@@ -29,8 +29,10 @@ import android.content.IntentFilter;
 import android.content.res.Configuration;
 import android.os.Build;
 
+import java.lang.ref.WeakReference;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.WeakHashMap;
 
 /**
  * 用于在plugin-loader中调用假的Application方法的接口
@@ -47,6 +49,22 @@ public class ShadowApplication extends ShadowContext {
             = new ShadowActivityLifecycleCallbacks.Holder();
 
     public boolean isCallOnCreate;
+
+    /**
+     * BroadcastReceiver到BroadcastReceiverWrapper对象到映射关系
+     * <p>
+     * 采用WeakHashMap<BroadcastReceiver, WeakReference<BroadcastReceiverWrapper>>
+     * 使key和value都采用弱引用持有，以保持原本BroadcastReceiver的GC回收时机。
+     * <p>
+     * BroadcastReceiver由原有业务代码强持有（也可能不持有），BroadcastReceiver原本在registerReceiver
+     * 之后交由系统持有，现在由BroadcastReceiverWrapper代替它被系统强持有。
+     * 所以BroadcastReceiverWrapper强引用持有BroadcastReceiver，保持了系统强引用BroadcastReceiver的关系。
+     * <p>
+     * 如果业务原本没有持有BroadcastReceiver，也就不会再有unregisterReceiver调用来，
+     * 也就不需要Map中有wrapper对应关系，所以用弱引用持有此关系没有影响。
+     */
+    final private Map<BroadcastReceiver, WeakReference<BroadcastReceiverWrapper>>
+            mReceiverWrapperMap = new WeakHashMap<>();
 
     @Override
     public Context getApplicationContext() {
@@ -175,5 +193,21 @@ public class ShadowApplication extends ShadowContext {
     @SuppressLint("NewApi")
     public static String getProcessName() {
         return Application.getProcessName();
+    }
+
+    public BroadcastReceiverWrapper receiverToWrapper(BroadcastReceiver receiver) {
+        if (receiver == null) {
+            return null;
+        }
+        synchronized (mReceiverWrapperMap) {
+            WeakReference<BroadcastReceiverWrapper> weakReference
+                    = mReceiverWrapperMap.get(receiver);
+            BroadcastReceiverWrapper wrapper = weakReference == null ? null : weakReference.get();
+            if (wrapper == null) {
+                wrapper = new BroadcastReceiverWrapper(receiver, this);
+                mReceiverWrapperMap.put(receiver, new WeakReference<>(wrapper));
+            }
+            return wrapper;
+        }
     }
 }

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowContext.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowContext.java
@@ -37,10 +37,6 @@ import android.view.LayoutInflater;
 import com.tencent.shadow.core.runtime.container.GeneratedHostActivityDelegator;
 import com.tencent.shadow.core.runtime.container.HostActivityDelegator;
 
-import java.lang.ref.WeakReference;
-import java.util.Map;
-import java.util.WeakHashMap;
-
 public class ShadowContext extends SubDirContextThemeWrapper {
     PluginComponentLauncher mPluginComponentLauncher;
     ClassLoader mPluginClassLoader;
@@ -51,21 +47,6 @@ public class ShadowContext extends SubDirContextThemeWrapper {
     ApplicationInfo mApplicationInfo;
     protected String mPartKey;
     private String mBusinessName;
-    /**
-     * BroadcastReceiver到BroadcastReceiverWrapper对象到映射关系
-     * <p>
-     * 采用WeakHashMap<BroadcastReceiver, WeakReference<BroadcastReceiverWrapper>>
-     * 使key和value都采用弱引用持有，以保持原本BroadcastReceiver的GC回收时机。
-     * <p>
-     * BroadcastReceiver由原有业务代码强持有（也可能不持有），BroadcastReceiver原本在registerReceiver
-     * 之后交由系统持有，现在由BroadcastReceiverWrapper代替它被系统强持有。
-     * 所以BroadcastReceiverWrapper强引用持有BroadcastReceiver，保持了系统强引用BroadcastReceiver的关系。
-     * <p>
-     * 如果业务原本没有持有BroadcastReceiver，也就不会再有unregisterReceiver调用来，
-     * 也就不需要Map中有wrapper对应关系，所以用弱引用持有此关系没有影响。
-     */
-    final private Map<BroadcastReceiver, WeakReference<BroadcastReceiverWrapper>>
-            mReceiverWrapperMap = new WeakHashMap<>();
 
     public ShadowContext() {
     }
@@ -303,18 +284,6 @@ public class ShadowContext extends SubDirContextThemeWrapper {
     }
 
     private BroadcastReceiverWrapper receiverToWrapper(BroadcastReceiver receiver) {
-        if (receiver == null) {
-            return null;
-        }
-        synchronized (mReceiverWrapperMap) {
-            WeakReference<BroadcastReceiverWrapper> weakReference
-                    = mReceiverWrapperMap.get(receiver);
-            BroadcastReceiverWrapper wrapper = weakReference == null ? null : weakReference.get();
-            if (wrapper == null) {
-                wrapper = new BroadcastReceiverWrapper(receiver, this);
-                mReceiverWrapperMap.put(receiver, new WeakReference<>(wrapper));
-            }
-            return wrapper;
-        }
+        return mShadowApplication.receiverToWrapper(receiver);
     }
 }

